Add ability to perform convergence analysis (#552)
open-AIMS/ADRIA.jl: docs/src/usage/analysis.md

![Plots of Temporal Sensitivities](/ADRIA.jl/dev/assets/imgs/tsa.png?raw=true "Temporal Sensitivity Analysis")

### Convergence Analysis for Sensitivity

> When undertaking sensitivity analysis it is importnant to have a sufficient sample size that the sensitivity measure has converged and is relatively stable. To understand convergence behaviour a 
> large sample can be drawn and split into smaller sub-samples. Plotting a sensitivity measure against increasing sample size can help to understand whether the sample size is sufficient for convergence.
> The function `convergence` can be used to calculated a sensitivity measure for an increasing number of samples. The result can then be plotted as band plots or a heat map using `viz.convergence`.

Reviewer: ```suggestion
When undertaking sensitivity analysis it is important to have sufficient number of samples such that the sensitivity measure converges to a stable state. To assess whether sufficient samples have been taken a convergence analysis can be conducted. One approach is to draw a large sample and then iteratively assess stability of the sensitivity metric using increasingly number of sub-samples. The sensitivity metric is described as having "converged" if there is little to no fluctuations/variance for a given sample size. The analysis can help determine if too little (or too many) samples have taken for the purpose of sensitivity analysis.

The function `convergence` can be used to calculated a sensitivity measure for an increasing number of samples. The result can then be plotted as band plots or a heat map using `viz.convergence`.
```

> The function `convergence` can be used to calculated a sensitivity measure for an increasing number of samples. The result can then be plotted as band plots or a heat map using `viz.convergence`.

Reviewer: You say `convergence` and later say `viz.convergence` to differentiate between the analysis and visualization methods.

Should the first be `sensitivity.convergence` or `analysis.convergence` instead?

```julia

> outcome = ADRIA.mean(ADRIA.metrics.scenario_total_cover(rs); dims=:timesteps)[timesteps=1]

Reviewer: Why `ADRIA.mean()` ? User should load the Statistics package themselves.

# Plot heat map of sensitivities for increasing number of scenarios for 
# individual factors (foi).
foi = [:dhw_scenario, :wave_scenario, :guided]
Si_N = convergence(scens, outcome, foi)

> ADRIA.viz.convergence(Si_N, Symbol.(foi))

Reviewer: No indication of where `convergence` comes from?

Is it not within a submodule?
Did you make it so it is not within a submodule and so it leaks into the global namespace?

Author: It's inside the ADRIA.sensitivity submodule. I'll add the .sensitivity at the front. I think in the example I copied from I imported the whole model so I didn't need to.

# Plot heat map of sensitivities for increasing number of scenarios for
# individual factors (the factors of interest, or "foi").
foi = [:dhw_scenario, :wave_scenario, :guided]
Si_N = ADRIA.sensitivity.convergence(scens, outcome, foi; opts=Dict(:viz_type=:heat_map))
ADRIA.viz.convergence(Si_N, Symbol.(foi))

# Plot heat map of sensitivities for increasing number of scenarios for
# particular model components.
components = [:EnvironmentalLayer, :Intervention]
Si_N = ADRIA.sensitivity.convergence(rs, scens, outcome, components; opts=Dict(:viz_type=:heat_map))
ADRIA.viz.convergence(Si_N, components)

```
![Plots of Convergence](/ADRIA.jl/dev/assets/imgs/colormap_convergence_factors.png?raw=true "Convergence Analysis- factors")
![Plots of Convergence](/ADRIA.jl/dev/assets/imgs/colormap_convergence_components.png?raw=true "Convergence Analysis- model components")

```julia
# Plot covergence as grid of band plots
Si_N = convergence(scens, outcome, foi)
ADRIA.viz.convergence(Si_N, foi, plot_overlay=false)

# Plot covergence as grid of overlayed band plots
ADRIA.viz.convergence(Si_N, foi, plot_overlay=true)

> ```

Reviewer: `plot_overlay` not a valid argument anymore?

![Plots of Temporal Sensitivities](/ADRIA.jl/dev/assets/imgs/tsa.png?raw=true "Temporal Sensitivity Analysis")

### Convergence Analysis

When undertaking sensitivity analysis it is important to have a sufficient number of samples
such that the sensitivity measure converges to a stable state. To assess whether sufficient
samples have been taken a convergence analysis can be conducted. One approach is to draw a
large sample and then iteratively assess stability of the sensitivity metric using an
increasing number of sub-samples. The sensitivity metric is described as having "converged"
if there is little to no fluctuations/variance for a given sample size. The analysis can
help determine if too little (or too many) samples have taken for the purpose of sensitivity
analysis.

The function `sensitivity.convergence` can be used to calculate a sensitivity measure for an
increasing number of samples. The result can then be plotted as band plots or a heat map
using `viz.convergence`.

```julia
using Statistics

outcome = dropdims(mean(ADRIA.metrics.scenario_total_cover(rs); dims=:timesteps), dims=:timesteps)

> # Display convergence for specific factors of interest ("foi") as heat map

Reviewer: ```suggestion
# Display convergence for specific factors of interest ("foi") as a heat map
```

foi = [:dhw_scenario, :wave_scenario, :guided]
Si_conv = ADRIA.sensitivity.convergence(scens, outcome, foi; opts=Dict(:viz_type=>:heatmap))
ADRIA.viz.convergence(Si_conv, foi)

> # Convergence analysis of factors grouped by model component as heat map

Reviewer: ```suggestion
# Convergence analysis of factors grouped by model component as a heat map
```

components = [:EnvironmentalLayer, :Intervention, :Coral]
Si_conv = ADRIA.sensitivity.convergence(rs, scens, outcome, components)
ADRIA.viz.convergence(Si_conv, components; opts=Dict(:viz_type=>:heatmap))
```

![Convergence analysis of factors](/ADRIA.jl/dev/assets/imgs/colormap_convergence_factors.png?raw=true "Convergence Analysis - factors")
![Grouped convergence analysis](/ADRIA.jl/dev/assets/imgs/colormap_convergence_components.png?raw=true "Convergence Analysis - model components")

```julia
# Display convergence plot within a single figure
ADRIA.viz.convergence(Si_conv, foi)

# Create a grid of figures for each factor of interest
ADRIA.viz.convergence(Si_conv, foi; opts=Dict(:plot_overlay=>false))

> ```

Reviewer: Please add note/warning about what the confidence interval represents here
